### ISSUE TITLE
construct_dput() ?

### ISSUE THREAD

**Issue by moodymudskipper** [cynkra/constructive#141]
* `dput()` constructs objects from internal type by adding attributes. Its flaws are that it's not always accurate, and that it doesn't use idiomatic constructors.
* `construct()` is accurate and uses idiomatic constructors

What if I want the accuracy only, because I want to explore the object ? (since str is a generic and that attributes can print funny)

We might (and should) have an "atomic" constructor for every class build on top of atomic vectors, and a "list" constructor for every  class built on top of lists (and so on for all internal types). This would allow this behaviour but it would still be tedious to set.

We might use a template that sets those, so we can do `construct(x, dput_template())`

OR (and?) we could have a construct_dput() function that is more restricted and uses the simplest constructor every time.

The first step in any case is to have the simple constructor for all, this means an `opt_*` function for all classes.

**Comment by moodymudskipper** [comment]
I'm not sure if I want the overhead of defining an opt function for every class just for this, it clutters the list of opts_ functions and adds boiler plate. At the same time this really is a class level option so maybe we don't have the choice.

**Comment by moodymudskipper** [comment]
What about this : 

* Every custom method must be associated to an `opts_?()` function

BUT

* `?construct()` mentions it and enumerates only those that have either : 
  * Other args then `constructor`
  * More than 1 value for `constructor` once we exclude the low level constructors "list" and "atomic"

In any case think `dput_template()` is not enough because it won't considered user defined classes.

So we need `construct_dput()`, or `construct_low_level()`, however, will always consider the "atomic" or "list" constructors whenever available,  and they should be available for every well designed user defined object built on top of atomic or list objects.

OR

We add an argument to `construct()` so that we consider by default the last constructor, assumed to be the atomic one, we have special casing for some of our constructors like "environment".  This way the opts functions can be used to use higher level constructors on top of simple defaults. `construct(x, opts_data.frame(), dput = TRUE)`, will use the data.frame constructor for data frames, but also tibbles. Not sure how to implement yet but that might be the right way to get at it

**Comment by moodymudskipper** [comment]
If we have a "next" constructor, falling back on next method, and a way to use next whenever available^, i.e everywhere except on the lowest constructors, for which we keep the default, we should have at the closest thing to a robust low level dput.

**Comment by moodymudskipper** [comment]
Let's generalise it a bit : 

* construct() chooses the highest level constructor as possible, so it means it's not stable, if the package evolves, some outputs might change
* construct_dput() chooses the lowest level, by default, but let's still allow passing opt functions to opt in for higher level construction, this way we can have a stable output that doesn't change when new constructors are added.

Since at this point we keep the exact same parameters, we might not need a new function, but only a boolean argument to define the bahavior, how would it be named ? `explicit = FALSE` ? `auto = TRUE` ?

Basically the `opts_*` functions would have their default mapped to `"next"` in those cases.
 
One corner case. If we fall back on the next constructor, should it be only the next that has been authorised ? e.g. if a tibble is corrupted and we have opted if for tibble but not data frame constructors, what happens ? Note that sometimes we use `NextMethod()` and sometimes we call a fallback constructor explicitly so this might lead to confusion.

**Comment by moodymudskipper** [comment]
There's a simple way to implement `construct_dput()`, that's not using dispatch at all and calling `.cstr_construct.default()`, that does the dispatch manually, something like : 

```
.cstr_construct <- function(x, ..., data = NULL, dput = FALSE) {
  data_name <- perfect_match(x, data)
  if (!is.null(data_name)) return(data_name)
  if (dput) return(.cstr_construct.default(x, ..., data = data, dput = TRUE))
  UseMethod(".cstr_construct")
}
```

We need to forward the arg however, and the most natural way is to provide it through construct and friends, then have a wrapper `construct_dput()`. We can also feed the info through the dots.

What I would also like however, is a way to restrict the classes flexibly, I want to be able to : 
* use only the simplest constructors (as in construct_dput)
* use only the default package constructors
* choose the packages or classes to include
* choose some packages or classes to exclude

Choosing the classes to include is convenient for snapshot stability, it doesn't guarantee that the package won't change or correct some output, but those are supposed to be changes for the better. New supported classes breaking snapshots however is more annoying.

We could have `classes = "all"` by default, and we can have "none" for the dput approach, and provide packages with curly braces, like "{ggplot2}". How do we exclude though ? We could use `-` like in tidy selection, where `-` is overloaded by `structure(x, names = "exclude")` so it's easy to postprocess.
